UnitTests/GitUITests/CommandsDialogs/CommitDialog/FormCommitTests.cs

```diff
@@ -0,0 +1,141 @@
+
+namespace GitUITests.CommandsDialogs.CommitDialog
+{
+    [Apartment(ApartmentState.STA)]
+    public class FormCommitTests
+    {
+        // Created once for the fixture
+        private ReferenceRepository _referenceRepository;
+
+        // Created once for each test
+        private GitUICommands _commands;
+
+        [SetUp]
+        public void SetUp()
+        {
+            if (_referenceRepository == null)
+            {
+                _referenceRepository = new ReferenceRepository();
+            }
+            else
+            {
+                _referenceRepository.Reset();
+            }
+
+            _commands = new GitUICommands(_referenceRepository.Module);
+        }
+
+        [TearDown]
+        public void TearDown()
+        {
+            _commands = null;
+        }
+
+        [OneTimeTearDown]
+        public void OneTimeTearDown()
+        {
+            _referenceRepository.Dispose();
+        }
+
+        [Test]
+        public void PreserveCommitMessageOnReopen()
+        {
+            var generatedCommitMessage = Guid.NewGuid().ToString();
+
+            RunFormCommitTest(formCommit =>
+            {
+                Assert.IsEmpty(formCommit.GetTestAccessor().Message.Text);
+                formCommit.GetTestAccessor().Message.Text = generatedCommitMessage;
+            });
+
+            RunFormCommitTest(formCommit =>
+            {
+                Assert.AreEqual(generatedCommitMessage, formCommit.GetTestAccessor().Message.Text);
+            });
+        }
+
+        [TestCase(CommitKind.Fixup)]
+        [TestCase(CommitKind.Squash)]
+        public void DoNotPreserveCommitMessageOnReopenFromSpecialCommit(CommitKind commitKind)
+        {
+            var generatedCommitMessage = Guid.NewGuid().ToString();
+
+            RunFormCommitTest(
+                formCommit =>
+                {
+                    string prefix = commitKind.ToString().ToLowerInvariant();
+                    Assert.AreEqual($"{prefix}! A commit message", formCommit.GetTestAccessor().Message.Text);
+                    formCommit.GetTestAccessor().Message.Text = generatedCommitMessage;
+                },
+                commitKind);
+
+            RunFormCommitTest(formCommit =>
+            {
+                Assert.IsEmpty(formCommit.GetTestAccessor().Message.Text);
+            });
+        }
+
+        [Test]
+        [Ignore("https://github.com/gitextensions/gitextensions/issues/4856")]
+        public void SelectMessageFromHistory()
+        {
+            var generatedCommitMessage = Guid.NewGuid().ToString();
+
+            RunFormCommitTest(formCommit =>
+            {
+                var commitMessageToolStripMenuItem = formCommit.GetTestAccessor().CommitMessageToolStripMenuItem;
+
+                // Verify the message appears correctly
+                commitMessageToolStripMenuItem.ShowDropDown();
+                Assert.AreEqual("A commit message...", commitMessageToolStripMenuItem.DropDownItems[0].Text);
```

Comment:
#4867 fixes the blank message bug, but you'll need to update the expected string to remove the ellipsis to make the test pass.